src/OpenRiaServices.Client.HttpDomainClient/Framework/BinaryHttpDomainClient.cs

```diff
@@ -0,0 +1,523 @@
+                    if (param.Value != null)
+                    {
+                        uriBuilder.Append(i++ == 0 ? '?' : '&');
+                        uriBuilder.Append(Uri.EscapeDataString(param.Key));
+                        uriBuilder.Append('=');
+                        var value = WebQueryStringConverter.ConvertValueToString(param.Value, param.Value.GetType());
+                        uriBuilder.Append(Uri.EscapeDataString(value));
+                    }
+                }
+            }
+
+            // Query options
+            if (queryOptions != null && queryOptions.Count > 0)
+            {
+                foreach (var queryPart in queryOptions)
+                {
+                    uriBuilder.Append(i++ == 0 ? "?$" : "&$");
+                    uriBuilder.Append(queryPart.QueryOperator);
+                    uriBuilder.Append("=");
+                    // Query strings seems to be double encoded
+                    uriBuilder.Append(Uri.EscapeDataString(Uri.EscapeDataString(queryPart.Expression)));
+                }
+            }
+
+            var uri = uriBuilder.ToString();
+            // TODO: Switch to POST if uri becomes to long, we can do so by returning nul ...l
+            if (uri.Length > 2048)
+                return null;
+            if (HttpClient.BaseAddress.OriginalString.Length
+                + operationName.Length
+                >= 2047)
+                return null;
+
+            return HttpClient.GetAsync(uri, DefaultHttpCompletionOption, cancellationToken);
+        }
+        #endregion
+
+        #region Private methods for reading responses
+
+        /// <summary>
+        /// Reads a response from the service and converts it to the specified return type.
+        /// </summary>
+        /// <param name="response">the <see cref="HttpResponseMessage"/> to deserialize</param>
+        /// <param name="operationName">name of operation invoked, used to verify returned xml</param>
+        /// <param name="returnType">Type which should be returned.</param>
+        /// <returns></returns>
+        /// <exception cref="OpenRiaServices.Client.DomainOperationException">On server errors which did not produce expected output</exception>
+        /// <exception cref="FaultException{DomainServiceFault}">If server returned a DomainServiceFault</exception>
+        private async Task<object> ReadResponseAsync(HttpResponseMessage response, string operationName, Type returnType)
+        {
+            // Always dispose using finally block below  respnse or we can leak connections
+            using (response)
+            {
+                // TODO: OpenRia 5.0 returns different status codes
+                // Need to read content and parse it even if status code is not 200
+                // It would make sens to one  check content type and only pase on msbin
+                if (!response.IsSuccessStatusCode && response.Content.Headers.ContentType.MediaType != "application/msbin1")
```

Comment:
@ehsangfl the extra check for media type should be somewhat similar to what you did in  https://github.com/OpenRIAServices/Samples/pull/7

do you think it will work well in this way, it works for all unit test but it was a long time ago I did tests against general HTTP error pages?